core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -343,12 +375,10 @@ private void logClusterHealthStateChange(ClusterStateHealth previousStateHealth,
 
     private boolean reroute(RoutingAllocation allocation) {
         boolean changed = false;
-        // first, clear from the shards any node id they used to belong to that is now dead
-        changed |= deassociateDeadNodes(allocation);
+        assert deassociateDeadNodes(allocation) == false : "dead nodes should be explicitly cleaned up. See deassociateDeadNodes";
 
-        // elect primaries *before* allocating unassigned, so backups of primaries that failed
-        // will be moved to primary state and not wait for primaries to be allocated and recovered (*from gateway*)
-        changed |= electPrimariesAndUnassignedDanglingReplicas(allocation);
+        assert electPrimariesAndUnassignedDanglingReplicas(allocation) == false:
```

Comment:
I was wondering why this assertion didn't trip when we cancel a primary shard using CancelAllocationCommand (we even have a test for that -> AllocationCommandTests.testCancelCommand, the last lines of the test). After investigation, I noticed that the `electPrimariesAndUnassignedDanglingReplicas` method does not correctly set the changed flag to true when `candidate != null` (it currently only does if `primarySwappedCandidate.relocatingNodeId() != null` or `IndexMetaData.isIndexUsingShadowReplicas(index.getSettings())` but it should do it always when `candidate != null`).